File: CollectionLayout-OC/CollectionLayout/ControllerGroup/PinterestViewController.m
//
//  PinterestViewController.m
//  PinterestDemo
//
//  Created by 苏沫离 on 2020/8/10.
//  Copyright © 2020 苏沫离. All rights reserved.
//
#define CellIdentifer @"PinterestCollectionCell"


#import "PinterestViewController.h"
#import "PinterestCollectionCell.h"
#import "YLCollectionPinterestLayout.h"
#import "CollectionSectionHeaderView.h"

@interface PinterestViewController ()
<UICollectionViewDelegate,UICollectionViewDataSource,
YLCollectionPinterestLayoutDelegate>
@property (nonatomic ,strong) UICollectionView *collectionView;

@end

@implementation PinterestViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = UIColor.whiteColor;
    [self.view addSubview:self.collectionView];
}

- (void)viewWillLayoutSubviews{
    [super viewWillLayoutSubviews];
    self.collectionView.frame = self.view.bounds;
}

#pragma mark - YLCollectionViewLayoutDelegate

-(CGSize)YLLayout:(YLCollectionPinterestLayout *)layout elementOfKind:(NSString *)elementOfKind referenceSizeInSection:(NSInteger)section{
    if ([elementOfKind isEqualToString:UICollectionElementKindSectionHeader]) {
        return CGSizeMake(CGRectGetWidth(UIScreen.mainScreen.bounds), 172 / 375.0 * CGRectGetWidth(UIScreen.mainScreen.bounds));
    }
    return CGSizeMake(CGRectGetWidth(UIScreen.mainScreen.bounds), 0.1);
}

- (CGFloat)YLLayout:(YLCollectionPinterestLayout *)layout heightForItemAtIndexPath:(NSIndexPath *)indexPath itemWidth:(CGFloat)itemWidth{
    DataModel *model = DataModel.shareDemoData[indexPath.row];
    return [model cellHeightByWidth:itemWidth];
}

#pragma mark - UICollectionViewDelegate

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView{
    return 2;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return DataModel.shareDemoData.count;
}

- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath{
    if ([kind isEqualToString:UICollectionElementKindSectionHeader]) {
        CollectionSectionHeaderView *headerView = [collectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:kCollectionSectionHeaderIdentifer forIndexPath:indexPath];
        return headerView;
    }
    CollectionSectionFooterView *footerView = [collectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionFooter withReuseIdentifier:kCollectionSectionFooterIdentifer forIndexPath:indexPath];
    return footerView;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    PinterestCollectionCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:CellIdentifer forIndexPath:indexPath];
    cell.model = DataModel.shareDemoData[indexPath.row];
    return cell;
}

#pragma mark - getter and setter

- (UICollectionView *)collectionView{
    if (_collectionView == nil) {
        YLCollectionPinterestLayout *layout = [[YLCollectionPinterestLayout alloc] init];
        CGFloat width = (CGRectGetWidth(UIScreen.mainScreen.bounds) - 16 * 2.0 - 12) / 2.0;
        layout.itemSize = CGSizeMake(width, 123 / 165.0 * width + 24);
        layout.delegate = self;

        UICollectionView *collectionView = [[UICollectionView alloc] initWithFrame:self.view.bounds collectionViewLayout:layout];
        collectionView.dataSource = self;
        collectionView.delegate = self;
        collectionView.showsVerticalScrollIndicator = NO;
        collectionView.showsHorizontalScrollIndicator = NO;
        [collectionView registerClass:CollectionSectionHeaderView.class forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:kCollectionSectionHeaderIdentifer];
        [collectionView registerClass:CollectionSectionFooterView.class forSupplementaryViewOfKind:UICollectionElementKindSectionFooter withReuseIdentifier:kCollectionSectionFooterIdentifer];
        [collectionView registerClass:PinterestCollectionCell.class forCellWithReuseIdentifier:CellIdentifer];
        collectionView.backgroundColor = UIColor.whiteColor;
        _collectionView = collectionView;
    }
    return _collectionView;
}

@end

File: CollectionLayout-Swift/CollectionLayout-Swift/ControllerGroup/PinterestViewController.swift
//
//  PinterestViewController.swift
//  CollectionLayout-Swift
//
//  Created by 苏沫离 on 2020/9/11.
//  Copyright © 2020 苏沫离. All rights reserved.
//
// 瀑布流：高度自适应

import UIKit

class PinterestViewController: UIViewController,UICollectionViewDelegate,
                               UICollectionViewDataSource, YLCollectionPinterestLayoutDelegate{
    ///懒加载
    lazy var collectionView : UICollectionView! = {
        let flowLayout = YLCollectionPinterestLayout()
        let width = (UIScreen.main.bounds.width - 16 * 2.0 - 12) / 2.0
        flowLayout.itemSize = CGSize(width: width, height: 123 / 165.0 * width + 24)
        flowLayout.delegate = self
        
         let collectionView = UICollectionView(frame: self.view.bounds, collectionViewLayout: flowLayout)
         collectionView.backgroundColor = UIColor.white
         collectionView.showsVerticalScrollIndicator = false
         collectionView.showsHorizontalScrollIndicator = false
         collectionView.delegate = self
         collectionView.dataSource = self
         collectionView.register(PinterestCollectionCell.self, forCellWithReuseIdentifier: "PinterestCollectionCell")
        collectionView.register(CollectionSectionHeaderView.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "HeaderView")
        collectionView.register(CollectionSectionFooterView.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionFooter, withReuseIdentifier: "footerView")
        return collectionView
    }()

    
    // MARK: - life cycyle
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.white
        self.view.addSubview(collectionView)
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        collectionView.frame = self.view.bounds
    }
    
    // MARK: - YLCollectionViewLayoutDelegate
    
    func YLLayoutReferenceSizeInSection(layout: YLCollectionPinterestLayout, elementOfKind: String, referenceSizeInSection: NSInteger) -> CGSize {
        if elementOfKind == UICollectionView.elementKindSectionHeader {
            return CGSize(width: UIScreen.main.bounds.width, height: 172 / 375.0 * UIScreen.main.bounds.width)
        }
        return CGSize(width: UIScreen.main.bounds.width, height: 0.1)
    }
    
    func YLLayoutHeightForItemAtIndexPath(layout: YLCollectionPinterestLayout, indexPath: NSIndexPath, itemWidth: CGFloat) -> CGFloat {
        return DataModel.shareDemoData()[indexPath.row].cellHeightByWidth(width: itemWidth)
    }
    
    // MARK: - UICollectionViewDelegate
    
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 2
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return DataModel.shareDemoData().count
    }
    
    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        if kind == UICollectionView.elementKindSectionHeader {
            let headerView = collectionView.dequeueReusableSupplementaryView(ofKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "HeaderView", for: indexPath)
            return headerView
        }
        let footerView = collectionView.dequeueReusableSupplementaryView(ofKind: UICollectionView.elementKindSectionFooter, withReuseIdentifier: "footerView", for: indexPath)
        return footerView
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
         let cell : PinterestCollectionCell = collectionView.dequeueReusableCell(withReuseIdentifier: "PinterestCollectionCell", for: indexPath) as! PinterestCollectionCell
         cell.model = DataModel.shareDemoData()[indexPath.row]
         return cell
    }
 }
